### PR TITLE
Fixed #33161 -- Enabled durability check for nested atomic blocks in TestCase.

### PR DESCRIPTION
**Pull request by kjagiello**
ticket-33161

### CODE REVIEW

File: django/db/transaction.py [hunk 1]
```diff
@@ -296,14 +305,14 @@ def __exit__(self, exc_type, exc_value, traceback):
                     connection.in_atomic_block = False
 
 
-def atomic(using=None, savepoint=True, durable=False):
+def atomic(using=None, savepoint=True, durable=False, _test_marker=False):
```
Comment (kjagiello) [inline django/db/transaction.py:308]: Probably not desirable to expose private arguments on public interfaces like this.

**Review by adamchainz — COMMENTED** [review]
After reviewing I came up with an alternative implementation: #14922

File: django/db/transaction.py [hunk 2]
```diff
@@ -207,9 +210,15 @@ def __enter__(self):
             connection.set_autocommit(False, force_begin_transaction_with_broken_autocommit=True)
             connection.in_atomic_block = True
 
+        if connection.in_atomic_block:
+            connection.atomic_blocks.append(self)
```
[the remark this answers is not in the document]
Comment (kjagiello) [reply django/db/transaction.py:214]: Good point! I'll see if we can somehow unify them.

File: django/test/testcases.py [hunk 3]
```diff
@@ -1146,7 +1146,7 @@ def _enter_atomics(cls):
         """Open atomic blocks for multiple databases."""
         atomics = {}
         for db_name in cls._databases_names():
-            atomics[db_name] = transaction.atomic(using=db_name)
+            atomics[db_name] = transaction.atomic(using=db_name, _test_marker=True)
```
[the remark this answers is not in the document]
Comment (kjagiello) [reply django/test/testcases.py:1149]: Seems like a much better approach! And yeah, I agree about the naming.

**Comment by adamchainz** [comment]
Did you see #14922 ?

**Comment by kjagiello** [comment]
@adamchainz Yep! Just saw it and I feel like it is a more straightforward solution to this specific ticket. 

Having the `atomic_blocks` stack would make sense if I would also change the way `connection.in_atomic_block` and `connection.savepoint_ids` work, but maybe it feels like an unnecessary undertaking.

**Comment by kjagiello** [comment]
Let's close this one and go with your PR instead

File: django/test/testcases.py [hunk 4]
```diff
@@ -1189,12 +1188,10 @@ def setUpClass(cls):
                 if value is not pre_attrs.get(name):
                     setattr(cls, name, TestData(name, value))
         except Exception:
-            transaction.Atomic._ensure_durability = True
             raise
```
Comment (Ian-Foote) [inline django/test/testcases.py:1191]: We don't need these two lines, or the starting `try` line.

**Review by blockchainGuru1018 — APPROVED** [review]

File: django/db/transaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/transaction.py:210]: We set `in_atomic_block` manually so it seems like we need all three :thinking: :

https://github.com/django/django/blob/adb4100e58d9ea073ee8caa454bb7c885b6a83ed/django/db/transaction.py#L189-L194

https://github.com/django/django/blob/adb4100e58d9ea073ee8caa454bb7c885b6a83ed/django/db/transaction.py#L216-L217

Potential cleanups are welcome, but we can process them separately.

File: tests/transactions/tests.py [hunk 5]
```diff
@@ -534,14 +534,7 @@ def test_nested_inner_durable(self):
...
 
-class DisableDurabiltityCheckTests(TestCase):
-    """
-    TestCase runs all tests in a transaction by default. Code using
-    durable=True would always fail when run from TestCase. This would mean
-    these tests would be forced to use the slower TransactionTestCase even when
-    not testing durability. For this reason, TestCase disables the durability
-    check.
-    """
+class DurableTests(TestCase):
```
Comment (felixxm) [inline tests/transactions/tests.py]: I noticed that we have the same tests for `DurableTests` and `DurableTransactionTests` so I added `DurableBaseTests`.

**Review by felixxm — COMMENTED** [review]
@kjagiello Thanks :+1: Welcome aboard :boat: 

I pushed final edits.

**Pull request merged**